File: leetcode/python/passedProblems/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if l1 is None:
            return l2;
        elif l2 is None:
            return l1;
        re = ListNode(0);
        p = re;
        add = 0;
        while(l1 is not None or l2 is not None):
            sum = add;
            if l1 is not None:
                sum += l1.val;
                l1 = l1.next;
            if l2 is not None:
                sum += l2.val;
                l2 = l2.next;
            add = sum / 10;
            sum %= 10;
            re.next = ListNode(sum);
            re = re.next;
        if add > 0:
            re.next = ListNode(add);
        return p.next;

File: leetcode/cpp/2-add-two-numbers.cpp
/*
 * 2个链表表示2整数，求和
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * h = new ListNode(0), *t;
        t = h;
        int sum = 0, add = 0;
        while(l1 != NULL || l2 != NULL){
            sum = 0;
            if(l1 != NULL) {sum += l1->val; l1 = l1->next;}
            if(l2 != NULL) {sum += l2->val; l2 = l2->next;}
            t->next = new ListNode((sum + add) % 10);
            add =(sum+add)/10;
            t = t->next;
        }
        if(add > 0) t->next = new ListNode(add);
        if(h->next == NULL) h->next = new ListNode(0);
        return h->next;

    }
};
